consumer-test/integrationTestController.sh

```diff
@@ -0,0 +1,138 @@
+    echo "[ERROR] ${message}" >&2
+    exit "${returnCode}"
+}
+
+function notify() {
+
+    local state=${1}
+    local description=${2}
+    local hash=${3}
+
+    echo "[INFO] Notifying about state \"${state}\" for commit \"${hash}\"."
+
+    curl -X POST \
+        --fail \
+        --silent \
+        --output /dev/null \
+        --data "{\"state\": \"${state}\", \"target_url\": \"${TRAVIS_BUILD_WEB_URL}\", \"description\": \"${description}\", \"context\": \"integration-tests\"}" \
+        --user "${INTEGRATION_TEST_VOTING_USER}:${INTEGRATION_TEST_VOTING_TOKEN}" \
+    "https://api.github.com/repos/SAP/jenkins-library/statuses/${hash}" || fail "Cannot send notification. curl return code: $?"
+}
+
+function cleanup() {
+    [ -z "${notificationThreadPid}" ] || kill -PIPE "${notificationThreadPid}" &>/dev/null
+}
+
+trap cleanup EXIT
+
+#
+# In case the build is performed for a pull request TRAVIS_COMMIT is a merge
+# commit between the base branch and the PR branch HEAD. That commit is actually built.
+# But for notifying about a build status we need the commit which is currenty
+# the HEAD of the PR branch.
+#
+# In case the build is performed for a simple branch (not associated with a PR)
+# In this case there is no merge commit between any base branch and HEAD of a PR branch.
+# The commit which we need for notifying about a build status is in this case simply
+# TRAVIS_COMMIT itself.
+#
+COMMIT_HASH_FOR_STATUS_NOTIFICATIONS="${TRAVIS_PULL_REQUEST_SHA}"
+[ -z "${COMMIT_HASH_FOR_STATUS_NOTIFICATIONS}" ] && COMMIT_HASH_FOR_STATUS_NOTIFICATIONS="${TRAVIS_COMMIT}"
+
+notify "pending" "Integration tests in progress." "${COMMIT_HASH_FOR_STATUS_NOTIFICATIONS}"
+
+WORKSPACES_ROOT=workspaces
+[ -e "${WORKSPACES_ROOT}"  ] && rm -rf ${WORKSPACES_ROOT}
+
+TEST_CASES=$(find testCases -name '*.yml')
+
+# This auxiliar thread is needed in order to produce some output while the
+# test are running. Otherwise the job will be canceled after 10 minutes without
+# output.
+while true; do sleep 10; echo "[INFO] Integration tests still running."; done &
+notificationThreadPid=$!
+
+declare -a processes
+i=0
+for f in ${TEST_CASES}
+do
+    testCase=$(basename "${f%.*}")
+    area=$(dirname "${f#*/}")
+    echo "[INFO] Running test case \"${testCase}\" in area \"${area}\"."
+    TEST_CASE_ROOT="${WORKSPACES_ROOT}/${area}/${testCase}"
+    [ -e "${TEST_CASE_ROOT}" ] && rm -rf "${TEST_CASE_ROOT}"
+    mkdir -p "${TEST_CASE_ROOT}" || fail "Cannot create test case root director for test case \"${testCase}\"." 1
+    source ./runTest.sh "${testCase}" "${TEST_CASE_ROOT}" &> "${TEST_CASE_ROOT}/log.txt" &
+    pid=$!
+    processes[$i]="${area}/${testCase}:${pid}"
+    echo "[INFO] Test case \"${testCase}\" in area \"${area}\" launched. (PID: \"${pid}\")."
+    let i=i+1
+done
+
+[ "${i}" == 0 ] && fail "No tests has been executed." 1
+
+#
+# wait for the test cases and cat the log
+for p in "${processes[@]}"
+do
+    area=$(dirname "${p%:*}")
```

Comment:
A lot of implicit assumption/mappings of dir names and (due to bash) a lot of cryptic code.  I would vote for better maintainable language in the future.

Reply:
>A lot of implicit assumption/mappings

Would be cool to know more about the implicit assumptions in order to make them more transparent/understandable. Basically we traverse a folder containing test config Each tests gets a dedicated folder for running the tests. That all sounds reasonable to me.

> (due to bash) a lot of cryptic code. I would vote for better maintainable language in the future.

Well, bash is the lingua franca. It works without more dependencies. Bash is present basically everywhere on operation systems relevant for our use cases (... now you can start hunting for contra dictionary examples, and of course you will find them, but beside some edge cases ...). IMO the code is not that cryptic, but might be a matter of taste.

I'm not agaist re-writing it in a language which is considered to be more convenient, e.g. groovy. The docu generator is also written in groovy.

The reason why I used bash is basically: the code before controlling the integration tests was also provided in bash.

But I would like to propose to start with this here and to improve over time.